Improve es_test decorator

joxl opened the pull request:
## Technical Summary

Changes the behavior the the `es_test` decorator when called with adapters.

1. Performs index cleanup via `add[Class]Cleanup()` method instead of `tearDown[Class]()`.
2. Can only perform setup/teardown on test case classes that implement an `add[Class]Cleanup()` method (the previous implementation did not require test case classes to comply with _any_ interface).
3. Is tolerant of an existing test index (now deletes and recreates rather than resulting in an Elastic exception).

Item :three: (7634eddc645) is temporary and will be reverted once tests are refactored to clean up after themselves.

Review: commit by commit.

## Safety Assurance

### Safety story

Tests only.

### Automated test coverage

Includes new test util tests.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

joxl committed:
Increase robustness of the `es_test` decorator

- use ``add[Class]Cleanup`` instead of `tearDown[Class]`` to ensure the
  cleanup logic gets called (even if `setUp[Class]` fails)

joxl committed:
User ``operation.reverse_run()`` instead of discrete delete.

joxl committed:
Perform a preemptive delete prior to creating test indexes

- necessary in order to tolerate existing, less courteous tests

joxl committed:
Only add cleanups for those that were successfully created

millerdev commented on corehq/apps/es/tests/utils.py:
```diff
@@ -125,23 +125,15 @@ def es_test_decorator(test):
-                yield adapter
-
-    adapters = list(requires)
+    comment = f"created for {test.__module__}.{test.__qualname__}"
```
Why use `__qualname__` instead of `__name__`?

I thought `__qualname__` included the module path, so was confused about why this is also adding `test.__module__`, but on testing in a shell I see that `TestCase.__qualname__` simply returns `"TestCase"`, the same as `TestCase.__name__`.

joxl replied:
I'm not sure I understand. Are you saying you're seeing "wrong" (index) comments, or is the second part of this (PR) comment confirming that you _don't_ see anything "wrong"?

millerdev replied:
I am just curious about why you're using `__qualname__` instead of `__name__` here?

It appears that they both return the same thing, so it should be fine. But I would have chosen `__name__` here since it seems more clear. I think `__qualname__` includes the module path in some contexts (maybe I'm wrong about that?), so it is confusing to see it used in a place that is also prepending the module path.

joxl replied:
Thanks for clarifying. `__qualname__` does not include the module path, but it does provide [more complete information](https://stackoverflow.com/a/58110139), which can be quite helpful with tests where it's not uncommon to have nested classes and such.

millerdev replied:
Ah, that makes sense. Thanks for clearing up my confusion.

millerdev commented on corehq/apps/es/tests/utils.py:
```diff
@@ -124,101 +125,119 @@ def es_test_decorator(test):
             return test(*args, **kw)
         finally:
-            teardown_func()
+            for operation in created:
```
Should this iterate in reverse so the cleanup happens in the revers order of setup? Maybe it doesn't matter in this case?
```suggestion
            for operation in reversed(created):
```

joxl replied:
Good point! In this case it doesn't matter (yet, anyhow), but your suggestion is the correct way to do it to prevent gotchas in the event it someday would matter.

joxl committed:
Clean up in decorated functions the same as TestCase classes

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

joxl committed:
Call `tearDown[Class]` before `do[Class]Cleanups`

- prevents poisoning future tests due to nose trying to handle an
  inconsistent test case class
